examples/petsc/multigrid.c

```diff
@@ -148,6 +173,44 @@ int main(int argc, char **argv) {
       dmorig = dmDist;
     }
   }
+  {
+    // Apply Kershaw transformation
+    Vec coord;
+    PetscInt ncoord;
+    PetscScalar *c;
+    DMGetCoordinatesLocal(dmorig, &coord);
+    VecGetLocalSize(coord, &ncoord);
+    VecGetArray(coord, &c);
+
+    for (PetscInt i = 0; i < ncoord; i += 3) {
+      PetscScalar x = c[i], y = c[i+1], z = c[i+2];
+      PetscInt layer = x*6;
+      PetscScalar lambda = (x-layer/6.0)*6;
+      c[i] = x;
+      switch (layer) {
+        case 0:
+          c[i+1] = left(eps, y);
+          c[i+2] = left(eps, z);
+          break;
+        case 1:
+        case 4:
+          c[i+1] = step(left(eps, y), right(eps, y), lambda);
+          c[i+2] = step(left(eps, z), right(eps, z), lambda);
+          break;
+        case 2:
+          c[i+1] = step(right(eps, y), left(eps, y), lambda/2);
+          c[i+2] = step(right(eps, z), left(eps, z), lambda/2);
+          break;
+        case 3:
+          c[i+1] = step(right(eps, y), left(eps, y), (1+lambda)/2);
+          c[i+2] = step(right(eps, z), left(eps, z), (1+lambda)/2);
+          break;
+        default:
+          c[i+1] = right(eps, y);
+          c[i+2] = right(eps, z);
+      }
+    }
+  }
```

Comment:
Could you please move this into a function up above rather than this block of code. (`main` is already really long and I don't think the implementation of this transform is essential to the narrative.)